### PR TITLE
[main] qwen-vl THD packed-sequence support and fixes

### PR DESCRIPTION
**Pull request by DAISY-gh**
## Summary
- Add Qwen-VL THD packed-sequence support path and related runtime/config fixes in Megatron-Bridge.
- Update Qwen-VL recipe/task-encoder/vlm-step flow for THD batch assembly and MRoPE/padding-mask handling.
- Megatron-LM dependency context: this work is aligned with upstream Open PR https://github.com/NVIDIA/Megatron-LM/pull/2645.
image
curves of SFT on cord_v2 dataset based on Qwen3.5 35BA3B.

## Release Notes

* **New Features**
  * Added batch-level sequence packing support for improved training efficiency in data processing pipelines.
  * Introduced new Qwen3.5-VL training recipes with Energon dataset integration and configurable packing options.

* **Improvements**
  * Enhanced robustness by making optional dependencies (modelopt, diffusion modules) gracefully degrade when unavailable.
  * Improved multi-modal model training with better position computation and token accounting for mixture-of-experts layers.
  * Extended data provider configuration with new shuffling and packing buffer parameters for flexible batch preparation.

### CODE REVIEW

File: src/megatron/bridge/models/conversion/auto_bridge.py [hunk 1]
```diff
@@ -51,6 +50,14 @@
 
 logger = logging.getLogger(__name__)
 
+try:
+    from modelopt.torch.quantization.utils import is_quantized
+except Exception:
+    # modelopt/scipy is optional for recipe import and training startup.
+    # Keep quantization-only export path available when dependency exists.
+    def is_quantized(_model: object) -> bool:
+        return False
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/models/conversion/auto_bridge.py:65]: Addressed in 7263f09b: narrowed the optional modelopt import guard to ImportError and added a warning log for fallback visibility.

File: src/megatron/bridge/recipes/__init__.py [hunk 2]
```diff
@@ -18,8 +18,14 @@
...
 
-from megatron.bridge.diffusion.recipes.flux.flux import *
-from megatron.bridge.diffusion.recipes.wan.wan import *
+# Diffusion recipes require optional dependencies (e.g. diffusers). Keep
+# non-diffusion training paths (LLM/VLM) usable when those deps are absent.
+try:
+    from megatron.bridge.diffusion.recipes.flux.flux import *
+    from megatron.bridge.diffusion.recipes.wan.wan import *
+except ModuleNotFoundError:
+    pass
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/__init__.py:31]: Addressed in 7263f09b: split optional diffusion imports into independent try/except blocks so one missing stack does not suppress the other.

File: src/megatron/bridge/recipes/qwen_vl/qwen3_vl.py [hunk 3]
```diff
@@ -23,9 +23,14 @@
...
 
 import torch
-from transformers import AutoTokenizer, Qwen3VLProcessor
+from transformers import AutoTokenizer
 from typing_extensions import TypedDict, Unpack
 
+try:
+    from transformers import Qwen3VLProcessor
+except Exception:
+    Qwen3VLProcessor = None
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/qwen_vl/qwen3_vl.py:40]: Addressed in 7263f09b: narrowed broad exception handling for Qwen3VLProcessor import/load and added warning logs before falling back to AutoProcessor.

File: src/megatron/bridge/recipes/qwen_vl/qwen35_vl.py [hunk 4]
```diff
@@ -385,6 +390,51 @@ def qwen35_vl_35b_a3b_sft_config(hf_path: str = "Qwen/Qwen3.5-35B-A3B") -> Confi
...
+    # Keep both packing modes disabled by default.
+    # - in-batch packing: dataset.pack_sequences_in_batch
+    # - THD batch-level packing: dataset.batch_level_packing
+    # Enable either mode explicitly via CLI override when needed.
+    cfg.dataset.pack_sequences_in_batch = False
+    cfg.dataset.batch_level_packing = False
+    # Default packing window; can still be overridden via CLI:
+    # dataset.packing_buffer_size=<N>
+    cfg.dataset.packing_buffer_size = 128
+    return cfg
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/qwen_vl/qwen35_vl.py:417]: Addressed in 7263f09b: updated the docstring to match actual behavior (packing disabled by default; enable via CLI override).

File: src/megatron/bridge/recipes/qwen_vl/data/energon/task_encoder.py [hunk 5]
```diff
@@ -339,66 +439,354 @@ def encode_sample(self, sample: ChatMLSample):
...
+            loss_mask[labels < 0] = 0.0
+
+            max_cu_len = max(len(cu) for cu in cu_lengths_list)
+            padded_cu = torch.zeros((batch_size, max_cu_len), dtype=torch.int32)
+            for i, cu in enumerate(cu_lengths_list):
+                padded_cu[i, : len(cu)] = cu
+
+            cu_seqlens_argmin_list = [len(cu) for cu in cu_lengths_list]
+            cu_seqlens_argmin = torch.tensor(cu_seqlens_argmin_list, dtype=torch.int32)
+            max_lengths = torch.tensor(max_lengths_list, dtype=torch.int32)
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/qwen_vl/data/energon/task_encoder.py:753]: Thanks for the suggestion. `cu_seqlens_argmin` is currently consumed via `.item()` / host-side scalar reads, where `torch.int32` and `torch.int64` are functionally equivalent. In this code path, we intentionally keep `int32` to stay consistent with surrounding packed-seq metadata tensors and avoid unnecessary dtype conversions. So here, changing to `int64` is a consistency preference, not a functional requirement. If we later standardize all sequence-length/index metadata to `int64` across the full pipeline, we can update this in one pass.

**Comment by yaoyu-33** [comment]
@DAISY-gh do you need this before MLM side pr merge in?

File: src/megatron/bridge/data/energon/energon_provider.py [hunk 6]
```diff
@@ -33,20 +34,43 @@ class EnergonProvider(DatasetProvider):
     num_workers: int_repr
     dataloader_type: str = "external"
     task_encoder: Optional[Any] = None
-    # Enable batch-level online sequence packing
+    # Legacy in-batch packing switch. Keep for compatibility with existing recipes.
```
Comment (yaoyu-33) [inline src/megatron/bridge/data/energon/energon_provider.py]: add deprecate msg, if people should not use this. Since updating this, we need to change our recipe default.

File: src/megatron/bridge/models/conversion/auto_bridge.py [hunk 7]
```diff
@@ -51,6 +50,19 @@
 
 logger = logging.getLogger(__name__)
 
+try:
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/auto_bridge.py:53]: why change?

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py [hunk 8]
```diff
@@ -49,6 +51,27 @@
 from megatron.bridge.models.qwen_vl.modelling_qwen3_vl.vision_model import Qwen3VLVisionModel
 
 
+logger = logging.getLogger(__name__)
+
+
+def _thd_diag_enabled() -> bool:
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py]: move to utils

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py [hunk 9]
```diff
@@ -49,6 +51,27 @@
...
+
+def _thd_diag_align_enabled() -> bool:
+    return os.environ.get("THD_DIAG_ALIGN", "0") not in ("0", "", "false", "False")
+
+
+def _thd_diag_mrope_enabled() -> bool:
+    return os.environ.get("THD_DIAG_MROPE", "0") not in ("0", "", "false", "False")
+
+
+def _rank0() -> bool:
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py]: i think we have is_rank_0

File: src/megatron/bridge/recipes/qwen_vl/qwen3_vl.py [hunk 10]
```diff
@@ -19,13 +19,23 @@
...
 from typing_extensions import TypedDict, Unpack
 
+try:
+    from transformers import Qwen3VLProcessor
+except (ImportError, ModuleNotFoundError) as exc:
+    logger.warning(
+        "Qwen3VLProcessor import unavailable; falling back to AutoProcessor: %s",
+        exc,
+    )
+    Qwen3VLProcessor = None
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/qwen_vl/qwen3_vl.py:40]: Good catch — fixed in 3e83d311. I moved `logger = logging.getLogger(__name__)` above the optional `Qwen3VLProcessor` import guard, so the fallback path no longer risks `NameError` when the import is unavailable.

File: src/megatron/bridge/recipes/qwen_vl/data/energon/task_encoder.py [hunk 11]
```diff
@@ -39,6 +41,67 @@
...
+        while True:
+            idx = _search_for_fit(sorted_sizes, remaining)
+            if idx == -1:
+                break
+            remaining -= sorted_sizes[idx]
+            sorted_sizes.pop(idx)
+            current_knapsack.append(sorted_samples.pop(idx))
+        knapsacks.append(current_knapsack)
+    return knapsacks
 def process_vision(
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/qwen_vl/data/energon/task_encoder.py:107]: Fixed in 3e83d311 — added the missing blank line between `greedy_knapsack` and `process_vision`.

File: src/megatron/bridge/recipes/qwen_vl/data/energon/task_encoder.py [hunk 12]
```diff
@@ -39,6 +41,67 @@
...
+        current_knapsack = []
+        remaining = max_capacity
+        while True:
+            idx = _search_for_fit(sorted_sizes, remaining)
+            if idx == -1:
+                break
+            remaining -= sorted_sizes[idx]
+            sorted_sizes.pop(idx)
+            current_knapsack.append(sorted_samples.pop(idx))
+        knapsacks.append(current_knapsack)
```
[the remark this answers is not in the document]
Comment (DAISY-gh) [reply src/megatron/bridge/recipes/qwen_vl/data/energon/task_encoder.py:103]: Added focused unit tests in 3e83d311 for `greedy_knapsack` covering: empty input, single item, exact-fill packing, and over-capacity `ValueError` path (`tests/unit_tests/recipes/qwen_vl/data/energon/test_task_encoder.py`).

**Comment by DAISY-gh** [comment]
> @DAISY-gh do you need this before MLM side pr merge in?

The Megatron-Bridge PR depends on the M-LM PR. Now the dev branch is ready and main branch is WIP.

File: src/megatron/bridge/data/energon/energon_provider.py (hunk 6 above)
Comment (DAISY-gh) [reply src/megatron/bridge/data/energon/energon_provider.py]: > add deprecate msg, if people should not use this. Since updating this, we need to change our recipe default.

I didn't remove the in-batch packing, just add a new packing called batch level packing. Let me rephrase the comment.

Comment (DAISY-gh) [reply src/megatron/bridge/data/energon/energon_provider.py]: Updated in 496b95b6. I clarified the EnergonProvider comments to explicitly distinguish the two modes: `pack_sequences_in_batch` packs within an already formed micro-batch, while `batch_level_packing` packs across a dataloader-side candidate buffer before micro-batch formation. Also removed wording that could imply immediate deprecation.

File: src/megatron/bridge/models/conversion/auto_bridge.py (hunk 7 above)
Comment (DAISY-gh) [reply src/megatron/bridge/models/conversion/auto_bridge.py:53]: Good question. The change intentionally narrows the optional-import fallback from broad `Exception` to `ImportError`, so we do not mask unrelated runtime errors while still keeping graceful fallback when modelopt/scipy is unavailable. This keeps the quantized-export path optional but makes real failures visible in logs.

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py (hunk 8 above)
Comment (DAISY-gh) [reply src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py]: Addressed in 89cfeda1. I moved the THD diagnostic/rank helpers into the existing `modelling_qwen3_vl/utils.py` (instead of a new helper file), and updated `model.py` to import from that shared utils module.

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py (hunk 9 above)
Comment (DAISY-gh) [reply src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/model.py]: Addressed in 89cfeda1. I introduced a shared `is_rank_0()` helper in `megatron.bridge.utils.common_utils` and switched Qwen3-VL diagnostics to use it, removing the local duplicate rank-0 implementation.

**Review by yaoyu-33 — DISMISSED** [review]

File: src/megatron/bridge/models/conversion/auto_bridge.py (hunk 7 above)
Comment (cuichenx) [reply src/megatron/bridge/models/conversion/auto_bridge.py:53]: is this still needed now that modelopt version is bumped?

**Comment by yaoyu-33** [comment]
defer to subsequent fixes for this part.
